File: android/sdk/ssdid-core/src/main/java/my/ssdid/sdk/domain/oid4vp/OpenId4VpHandler.kt
package my.ssdid.sdk.domain.oid4vp

import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonPrimitive
import my.ssdid.sdk.domain.vault.Vault

class NoMatchingCredentialsException(message: String) : Exception(message)

data class PresentationReviewResult(
    val authRequest: AuthorizationRequest,
    val matches: List<MatchResult>
)

class OpenId4VpHandler(
    private val transport: OpenId4VpTransport,
    private val peMatcher: PresentationDefinitionMatcher,
    private val dcqlMatcher: DcqlMatcher,
    private val vault: Vault
) {

    suspend fun processRequest(uri: String): Result<PresentationReviewResult> = runCatching {
        val parsed = AuthorizationRequest.parse(uri).getOrThrow()

        val authRequest = if (parsed.requestUri != null) {
            val json = transport.fetchRequestObject(parsed.requestUri)
            AuthorizationRequest.parseJson(json).getOrThrow()
        } else {
            parsed
        }

        val storedVcs = vault.listStoredSdJwtVcs()

        val matches = when {
            authRequest.presentationDefinition != null ->
                peMatcher.match(authRequest.presentationDefinition, storedVcs)
            authRequest.dcqlQuery != null ->
                dcqlMatcher.match(authRequest.dcqlQuery, storedVcs)
            else -> emptyList()
        }

        if (matches.isEmpty()) {
            authRequest.responseUri?.let { responseUri ->
                runCatching { transport.postError(responseUri, "access_denied", authRequest.state) }
            }
            throw NoMatchingCredentialsException("No stored credentials match the request")
        }

        PresentationReviewResult(authRequest, matches)
    }

    suspend fun submitPresentation(
        authRequest: AuthorizationRequest,
        matchResult: MatchResult,
        selectedClaims: List<String>,
        algorithm: String,
        signer: (ByteArray) -> ByteArray
    ): Result<Unit> = runCatching {
        val responseUri = authRequest.responseUri
            ?: throw IllegalStateException("No response_uri in authorization request")
        val nonce = authRequest.nonce
            ?: throw IllegalStateException("No nonce in authorization request")

        val vpToken = VpTokenBuilder.build(
            storedSdJwtVc = matchResult.credential,
            selectedClaims = selectedClaims,
            audience = authRequest.clientId,
            nonce = nonce,
            algorithm = algorithm,
            signer = signer
        )

        val definitionId = when {
            authRequest.presentationDefinition != null ->
                authRequest.presentationDefinition["id"]?.jsonPrimitive?.contentOrNull
                    ?: throw IllegalStateException("Missing presentation_definition id")
            authRequest.dcqlQuery != null ->
                matchResult.descriptorId  // DCQL uses the credential query id
            else -> throw IllegalStateException("No query type in authorization request")
        }

        val submission = PresentationSubmission.create(
            definitionId = definitionId,
            descriptorIds = listOf(matchResult.descriptorId)
        )

        transport.postVpResponse(responseUri, vpToken, submission.toJson(), authRequest.state)
    }
}

File: sdk/ios/SsdidCore/Sources/SsdidCore/Domain/OpenId4Vp/OpenId4VpHandler.swift
import Foundation

/// Result of processing an authorization request before user consent.
public struct PresentationReviewResult {
    public let authRequest: AuthorizationRequest
    public let matches: [VpMatchResult]

    public init(authRequest: AuthorizationRequest, matches: [VpMatchResult]) {
        self.authRequest = authRequest
        self.matches = matches
    }
}

/// Protocol for listing stored SD-JWT VCs, implemented by VaultStorage.
public protocol SdJwtVcStore {
    func listSdJwtVcs() async -> [StoredSdJwtVc]
}

/// Orchestrates the OpenID4VP presentation flow.
public final class OpenId4VpHandler: @unchecked Sendable {

    private let transport: OpenId4VpTransport
    private let peMatcher: PresentationDefinitionMatcher
    private let dcqlMatcher: DcqlMatcher
    private let vcStore: SdJwtVcStore

    public     init(
        transport: OpenId4VpTransport,
        peMatcher: PresentationDefinitionMatcher = PresentationDefinitionMatcher(),
        dcqlMatcher: DcqlMatcher = DcqlMatcher(),
        vcStore: SdJwtVcStore
    ) {
        self.transport = transport
        self.peMatcher = peMatcher
        self.dcqlMatcher = dcqlMatcher
        self.vcStore = vcStore
    }

    /// Processes an authorization request URI: parses, resolves request_uri if present,
    /// matches credentials, and returns matches for user review.
    public     func processRequest(uri: String) async throws -> PresentationReviewResult {
        let parsed = try AuthorizationRequest.parse(uri)

        let authRequest: AuthorizationRequest
        if let requestUri = parsed.requestUri {
            let json = try await transport.fetchRequestObject(requestUri: requestUri)
            authRequest = try AuthorizationRequest.parseJson(json)
        } else {
            authRequest = parsed
        }

        let storedVcs = await vcStore.listSdJwtVcs()

        let matches: [VpMatchResult]
        if let pd = authRequest.presentationDefinition {
            matches = peMatcher.match(pd: pd, credentials: storedVcs)
        } else if let dcql = authRequest.dcqlQuery {
            matches = dcqlMatcher.match(dcql: dcql, credentials: storedVcs)
        } else {
            matches = []
        }

        if matches.isEmpty {
            if let responseUri = authRequest.responseUri {
                try? await transport.postError(
                    responseUri: responseUri,
                    error: "access_denied",
                    state: authRequest.state
                )
            }
            throw OpenId4VpError.noMatchingCredentials("No stored credentials match the request")
        }

        return PresentationReviewResult(authRequest: authRequest, matches: matches)
    }

    /// Submits a VP presentation after user consent.
    public     func submitPresentation(
        authRequest: AuthorizationRequest,
        matchResult: VpMatchResult,
        selectedClaims: [String],
        algorithm: String,
        signer: @Sendable (Data) -> Data
    ) async throws {
        guard let responseUri = authRequest.responseUri else {
            throw OpenId4VpError.invalidRequest("No response_uri in authorization request")
        }
        guard let nonce = authRequest.nonce else {
            throw OpenId4VpError.invalidRequest("No nonce in authorization request")
        }

        let vpToken = try VpTokenBuilder.build(
            storedSdJwtVc: matchResult.credential,
            selectedClaims: selectedClaims,
            audience: authRequest.clientId,
            nonce: nonce,
            algorithm: algorithm,
            signer: signer
        )

        // Determine definition ID from PE or DCQL query
        let definitionId: String
        if let pd = authRequest.presentationDefinition,
           let pdId = pd["id"] as? String {
            definitionId = pdId
        } else if authRequest.dcqlQuery != nil {
            definitionId = matchResult.descriptorId
        } else {
            throw OpenId4VpError.invalidRequest("Missing presentation_definition id or dcql_query")
        }

        let submission = PresentationSubmission.create(
            definitionId: definitionId,
            descriptorIds: [matchResult.descriptorId]
        )

        try await transport.postVpResponse(
            responseUri: responseUri,
            vpToken: vpToken,
            presentationSubmission: try submission.toJson(),
            state: authRequest.state
        )
    }
}
